### PR TITLE
PM Blocks Menu Implementation (#FOUR-7370)

### PR DESCRIPTION
**Pull request by sanjacornelius**
This pull request introduces the PM Blocks menu feature, addressing ticket [FOUR-7370](https://processmaker.atlassian.net/browse/FOUR-7370). The purpose of this feature is to enhance the Modeler tab by incorporating the display and utilization of PM Blocks. PM Blocks can now be seamlessly integrated within the modeler and executed as part of a process.

## Changes:

- Introduces a new method, `generateExports`, responsible for initiating an API call to the database. This call retrieves the stored PM Blocks and processes the response. The method iterates over the retrieved data and configures the corresponding node object. The resulting PM Blocks are then exported and stored in an array.

- Implements a new method called `registerPmBlocks`, which ensures the proper registration of PM Blocks within the modeler. This method is triggered during the initialization of the modeler, enabling seamless integration of PM Blocks functionality.

## To Test:

1. Ensure that the [PM Blocks](https://github.com/ProcessMaker/package-pm-blocks/pull/3) package branch  is `feature/FOUR-7370` installed.
2. Verify that PM Blocks are properly configured in the application.
3. Create a new process within the Modeler.
4. Check that the PM Block tab is visible in the Modeler interface.
5. Click on the PM Block tab to ensure it opens successfully and displays the list of available PM Blocks.
6. Click and drag a PM Block from the PM Block tab onto the modeler BPMN canvas.
7. Confirm that the dropped PM Block is visually represented on the canvas and can be interacted with (e.g., moving, resizing, connecting with other elements).
8. Run the process that includes the PM Block.
9. Validate that the process executes the associated PM Block properly when triggered during the process flow.
10. Verify that the PM Block functions as expected, performing the intended actions within the process.
11. Perform additional tests with different PM Blocks, covering various scenarios and functionalities.

## Related Tickets & Packages
-  PM Block PR - https://github.com/ProcessMaker/package-pm-blocks/pull/3
- Ticket: [FOUR-7370](https://processmaker.atlassian.net/browse/FOUR-7370)

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

[FOUR-7370]: https://processmaker.atlassian.net/browse/FOUR-7370?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[FOUR-7370]: https://processmaker.atlassian.net/browse/FOUR-7370?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/rails/explorer-rail/explorer.vue [hunk 1]
```diff
@@ -84,7 +90,10 @@ export default {
         v-on="$listeners"
       />
     </div>
-    <div class="pm-blocks__container">
+    <div class="pm-blocks__container"  v-if="tabIndex === 1">
+      <pm-blocks-loop
+        v-on="$listeners"
+      />
       <!--   Here goes the PM Blocks   -->
```
Comment (josechirivella) [inline src/components/rails/explorer-rail/explorer.vue:97]: remove this comment

File: src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        height: '40px',
+        width: '40px',
+        zIndex: '10',
+        opacity: '0.5',
+        pointerEvents: 'none',
+      },
+    };
+  },
+  methods: {
+    onClickHandler(event, control) {
```
Comment (josechirivella) [inline src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue:26]: Wasn't this the code you just refactored @TinCodes ?

File: src/nodeTypesStore.js [hunk 3]
```diff
@@ -51,10 +54,41 @@ export default new Vuex.Store({
...
     clearFilteredNodes(state) {
       state.filteredNodeTypes = [];
       state.searchTerm = '';
     },
+    clearFilteredPmBlockNodes(state) {
+      // TODO: Clear Pm Block search bar
+      console.log('CLEAR FILTER PM BLOCK NODE', state);
+      // state.filteredNodeTypes = [];
+      // state.searchTerm = '';
+    },
```
Comment (josechirivella) [inline src/nodeTypesStore.js:91]: I would add some similar logic or maybe reuse the component I created for the searchBar and passing like a SeachBarType to switch between regular nodeTypes and PmBlocks. That way we can do some v-if in rendering or logic.
Then on switching tabs, I would call here in this function `state.filteredNodeTypes = []` to clear the search bar to 0

File: src/nodeTypesStore.js [hunk 4]
```diff
@@ -51,10 +54,41 @@ export default new Vuex.Store({
...
+    setFilteredPmBlockNodeTypes(state, searchTerm) {
+      // TODO: Configure Pm Block search bar
+      console.log('FILTER PM BLOCK NODE TYPES', state, searchTerm);
+      // const nodeTypes = state.nodeTypes;
+      // state.searchTerm = searchTerm;
+      // const allNodes = uniqBy([...pinnedNodeTypes, ...nodeTypes], 'id');
+      // state.filteredNodeTypes = allNodes.filter(node => {
+      //   return node.label.toLowerCase().includes(searchTerm.toLowerCase());
+      // });
+    },
```
Comment (josechirivella) [inline src/nodeTypesStore.js:81]: same comment here. To replicate the same functionality from the search bar functionality to the pmBlocks, maybe we can have just one function to handle both scenarios to clear the searchTerm and searchBar

File: src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    searchTerm() {
+      return nodeTypesStore.getters.getSearchTerm;
+    },
+  },
+};
+</script>
+
+<template>
+  <div id="pmBlockNodeTypesList">
+    <!-- <div id="filteredPmBlockNodes-container" v-if="filteredPmBlockNodes.length > 0">
```
Comment (josechirivella) [inline src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue:100]: you are getting somewhere here but let me know . This component was hard to write, so much logic happening

File: .eslintrc.js [hunk 6]
```diff
@@ -12,6 +12,7 @@ module.exports = {
 
   rules: {
     semi: ['error', 'always'],
+    'no-console': 'off',
```
Comment (josechirivella) [inline .eslintrc.js:15]: I'm going to add a comment here to remove this. What you can do for eslint not to bother you is in the `vue.config.js`, add a line as `lintOnSave: false` and you are good to go

**Review by josechirivella — CHANGES_REQUESTED** [review]

File: src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue (hunk 2 above)
Comment (TinCodes) [reply src/components/rails/explorer-rail/pmBlocksLoop/pmBlocksLoop.vue:26]: Yes it is, there is now a new Mixin (`src/mixins/clickAndDrop.js`) that should handle the drag and drop. The updated file `src/components/rails/explorer-rail/nodeTypesLoop/nodeTypesLoop.vue` should be able to provide an example of how it is being used.

**Review by mcraeteisha — APPROVED** [review]
Approved!

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-f4b50bc04a.eng.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed 2023-06-23T19:41:55+00:00

**Comment by pmkerberos** [comment]
QA server was successfully deployed 2023-06-23T20:07:40+00:00

**Pull request merged**